File: backend/src/test/java/org/globe42/web/security/AuthenticationControllerTest.java
package org.globe42.web.security;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

import java.util.Optional;

import org.globe42.dao.UserDao;
import org.globe42.domain.User;
import org.globe42.test.BaseTest;
import org.globe42.web.exception.UnauthorizedException;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

/**
 * Unit tests for AuthenticationController
 * @author JB Nizet
 */
public class AuthenticationControllerTest extends BaseTest {
    @Mock
    private UserDao mockUserDao;

    @Mock
    private PasswordDigester mockPasswordDigester;

    @Mock
    private JwtHelper mockJwtHelper;

    @InjectMocks
    private AuthenticationController controller;

    @Test(expected = UnauthorizedException.class)
    public void shouldThrowWhenUnknownUser() {
        CredentialsDTO credentials = createCredentials();

        when(mockUserDao.findNotDeletedByLogin(credentials.getLogin())).thenReturn(Optional.empty());

        controller.authenticate(credentials);
    }

    @Test(expected = UnauthorizedException.class)
    public void shouldThrowWhenPasswordDoesntMatch() {
        CredentialsDTO credentials = createCredentials();

        User user = createUser();
        when(mockUserDao.findNotDeletedByLogin(credentials.getLogin())).thenReturn(Optional.of(user));
        when(mockPasswordDigester.match(credentials.getPassword(), user.getPassword())).thenReturn(false);

        controller.authenticate(credentials);
    }

    @Test
    public void shouldAuthenticate() {
        CredentialsDTO credentials = createCredentials();

        User user = createUser();
        when(mockUserDao.findNotDeletedByLogin(credentials.getLogin())).thenReturn(Optional.of(user));
        when(mockPasswordDigester.match(credentials.getPassword(), user.getPassword())).thenReturn(true);
        String token = "token";
        when(mockJwtHelper.buildToken(user.getId())).thenReturn(token);
        AuthenticatedUserDTO result = controller.authenticate(credentials);

        assertThat(result.getId()).isEqualTo(user.getId());
        assertThat(result.getLogin()).isEqualTo(user.getLogin());
        assertThat(result.isAdmin()).isEqualTo(user.isAdmin());
        assertThat(result.getToken()).isEqualTo(token);
    }

    public static CredentialsDTO createCredentials() {
        return new CredentialsDTO("JB", "passw0rd");
    }

    public static User createUser() {
        User user = new User(1L);
        user.setLogin("JB");
        user.setPassword("hashedPassword");
        user.setAdmin(true);
        return user;
    }
}

File: backend/src/test/kotlin/org/globe42/web/security/AuthenticationControllerTest.kt
package org.globe42.web.security

import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.globe42.dao.UserDao
import org.globe42.domain.User
import org.globe42.web.exception.UnauthorizedException
import org.junit.jupiter.api.Test

/**
 * Unit tests for AuthenticationController
 * @author JB Nizet
 */
class AuthenticationControllerTest {
    private val mockUserDao = mockk<UserDao>()

    private val mockPasswordDigester = mockk<PasswordDigester>()

    private val mockJwtHelper = mockk<JwtHelper>()

    private val controller = AuthenticationController(mockUserDao, mockPasswordDigester, mockJwtHelper)

    @Test
    fun `should throw when unknown user`() {
        val credentials = createCredentials()

        every { mockUserDao.findNotDeletedByLogin(credentials.login) } returns null

        assertThatExceptionOfType(UnauthorizedException::class.java).isThrownBy { controller.authenticate(credentials) }
    }

    @Test
    fun `should throw when password doesnt match`() {
        val credentials = createCredentials()

        val user = createUser()
        every { mockUserDao.findNotDeletedByLogin(credentials.login) } returns user
        every { mockPasswordDigester.match(credentials.password, user.password) } returns false

        assertThatExceptionOfType(UnauthorizedException::class.java).isThrownBy { controller.authenticate(credentials) }
    }

    @Test
    fun `should authenticate`() {
        val credentials = createCredentials()

        val user = createUser()
        every { mockUserDao.findNotDeletedByLogin(credentials.login) } returns user
        every { mockPasswordDigester.match(credentials.password, user.password) } returns true
        val token = "token"
        every { mockJwtHelper.buildToken(user.id) } returns token
        val result = controller.authenticate(credentials)

        assertThat(result.id).isEqualTo(user.id)
        assertThat(result.login).isEqualTo(user.login)
        assertThat(result.admin).isEqualTo(user.admin)
        assertThat(result.token).isEqualTo(token)
    }
}

internal fun createCredentials(): CredentialsCommandDTO {
    return CredentialsCommandDTO("JB", "passw0rd")
}

internal fun createUser(): User {
    val user = User(1L)
    user.login = "JB"
    user.password = "hashedPassword"
    user.admin = true
    return user
}
